pkg/resources/management.cattle.io/v3/project/validator.go

```diff
@@ -0,0 +1,211 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kubernetes/pkg/kubelet/util/format"
+	"k8s.io/utils/trace"
+)
+
+const (
+	systemProjectLabel = "authz.management.cattle.io/system-project"
+)
+
+// Validator implements admission.ValidatingAdmissionWebhook.
+type Validator struct {
+	admitter admitter
+}
+
+// NewValidator returns a project validator.
+func NewValidator(namespaceCache corev1controllers.NamespaceCache) *Validator {
+	return &Validator{
+		admitter{
+			namespaceCache: namespaceCache,
+		},
+	}
+}
+
+// GVR returns the GroupVersionKind for this CRD.
+func (v *Validator) GVR() schema.GroupVersionResource {
+	return gvr
+}
+
+// Operations returns list of operations handled by this validator.
+func (v *Validator) Operations() []admissionregistrationv1.OperationType {
+	return []admissionregistrationv1.OperationType{
+		admissionregistrationv1.Create,
+		admissionregistrationv1.Update,
+		admissionregistrationv1.Delete,
+	}
+}
+
+// ValidatingWebhook returns the ValidatingWebhook used for this CRD.
+func (v *Validator) ValidatingWebhook(clientConfig admissionregistrationv1.WebhookClientConfig) []admissionregistrationv1.ValidatingWebhook {
+	validatingWebhook := admission.NewDefaultValidatingWebhook(v, clientConfig, admissionregistrationv1.NamespacedScope, v.Operations())
+	return []admissionregistrationv1.ValidatingWebhook{*validatingWebhook}
+}
+
+// Admitters returns the admitter objects used to validate secrets.
+func (v *Validator) Admitters() []admission.Admitter {
+	return []admission.Admitter{&v.admitter}
+}
+
+type admitter struct {
+	namespaceCache corev1controllers.NamespaceCache
+}
+
+// Admit handles the webhook admission request sent to this webhook.
+func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResponse, error) {
+	listTrace := trace.New("project Admit", trace.Field{Key: "user", Value: request.UserInfo.Username})
+	defer listTrace.LogIfLong(admission.SlowTraceDuration)
+
+	oldProject, newProject, err := objectsv3.ProjectOldAndNewFromRequest(&request.AdmissionRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get old and new projects from request: %w", err)
+	}
+
+	if request.Operation == admissionv1.Delete {
+		return a.admitDelete(oldProject)
+	}
+	return a.admitCreateOrUpdate(oldProject, newProject)
+}
+
+func (a *admitter) admitDelete(project *v3.Project) (*admissionv1.AdmissionResponse, error) {
+	if project.Labels[systemProjectLabel] == "true" {
+		return admission.ResponseBadRequest("System Project cannot be deleted"), nil
+	}
+	return admission.ResponseAllowed(), nil
+}
+
+func (a *admitter) admitCreateOrUpdate(oldProject, newProject *v3.Project) (*admissionv1.AdmissionResponse, error) {
+	projectQuota := newProject.Spec.ResourceQuota
+	nsQuota := newProject.Spec.NamespaceDefaultResourceQuota
+	if projectQuota == nil && nsQuota == nil {
+		return admission.ResponseAllowed(), nil
+	}
+	allowed, msg, err := checkQuotaFields(projectQuota, nsQuota)
```

Comment:
We'd have to check both allowed and err at the same time to  figure out whether err was an error or an invalid request. Splitting it splits the concerns so that each return value can have just one easily discernible meaning. The other way I thought of doing it was just returning msg and err and checking if msg was empty, but this seemed more readable to me.

Reply:
Right, you may not need `allowed`, since in the function, when allowed is true, msg is always "", and err is always nil. So you could make it consistent with `checkQuotaValues` that only returns an error. Three returned values just looks complex.

I'd say using errors.Is or As to inspect the error is OK, we do it in a few places in this repo, but it's up to you...

Reply:
> I'd say using errors.Is or As to inspect the error is OK, we do it in a few places in this repo, 

It's not really an error, it's a successful negative response. I'd rather not have to interpret one of two different semantics for `error`, `error` should only mean something unexpected went wrong.

Reply:
I would tend to agree with @maxsokolovsky here for a couple of reasons:
- First, this is an established pattern in libraries (try to validate, error if we can't) that we use, like [this one](https://github.com/kubernetes/kubernetes/blob/1620473a9a01dd6bbef3398c0acb2e581d0a13c3/pkg/registry/rbac/validation/rule.go#L53) from upstream
- Second, this pattern (try to validate, err if we can't) is in use in [this file](https://github.com/rancher/webhook/blob/c84166079d2f204a3fd13482005c2df8dd35d82b/pkg/resources/management.cattle.io/v3/project/validator.go#L168). Namely, the `namespaceQuotaFits` fucntion will return an error if it confirms that the namespace quota doesn't fit - it was able to confirm it doesn't work, but it returned an error anyways.

This being said, I think the code is clear enough to follow so this one is more of implementer's choice.

Reply:
> First, this is an established pattern in libraries

It's not the pattern in webhook though - `Admit` makes a distinction between an `AdmissionResponse` and an error. If we wanted to make this consistent, then `Admit` should return only an error and `admission.NewValidatingHandlerFunc` should do the work of parsing it into a response.

> Second, this pattern (try to validate, err if we can't) is in use in this file

Valid point that my choice to wrap the `QuotaFits` returned bool in an error is inconsistent with the choice made here, will rework to at least make them consistent with each other.

Reply:
I found some upstream examples where the validation error and unexpected error are returned separately https://github.com/kubernetes/apiserver/blob/cc544e7bf1c20d41f4f61688a91ff9681d58693c/pkg/registry/rest/update.go#L87 - so even upstream is inconsistent about this.

I changed these helper methods to return a field.Error and error, instead of the bool, string, and error, so these should all be handling errors in a consistent way now.